DM-42188: Make RemoteButler usable from services

dhirving opened the pull request:
## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

dhirving committed:
Decouple Butler constructor params from subclass

Decoupled the set of parameter passed to Butler() from those passed to its subclasses.  RemoteButler construction will be changing to require injection of some long-lived objects, so it needs a different constructor signature than the original Butler constructor.

There are several parameters to Butler like "collections" and "run" which control the behavior of a Butler instance and are common to both DirectButler and RemoteButler.  Upcoming changes to Butler construction will require these to be passed around through multiple functions.  To avoid repeating them they are pulled out to a dataclass object.

dhirving committed:
Add a factory for instantiating RemoteButlers

To provide a place for storing state that is shared between multiple instances of RemoteButler, add a RemoteButlerFactory class that can be used to get instances of RemoteButler.

This also simplifies the constructor logic for both Butler types by making Butler.from_config() the only place responsible for knowing about the configuration search logic.

dhirving committed:
Split Butler "copy constructor" into new function

The DirectButler constructor used to have two forms that were logically separate but combined into one interface:
1. Standard new instance creation
2. Copy construction

This became more awkward when part of the construction process was moved to Butler.from_config, because 'config' and 'butler' are mutually exclusive options but we always need to instantiate a config to find out which class to load.

Moved the copy construction form into a separate instance method _clone().  This will likely become public at some point in the near future.

dhirving committed:
Make ButlerRepoIndex threadsafe

Python does not guarantee that dict mutations are threadsafe, and in particular the current CPython implementation is not if __eq__ is defined for the object used as the key.  (Almost every Butler-related object defines a custom __eq__.)

dhirving committed:
Move DirectButler.__new__ to a factory function

Moved most of the logic from DirectButler.__new__ into a factory function DirectButler.create_from_config.
 This allows us to use  __new__ as a common constructor shared by create_from_config() and clone().  This will help ensure that any new instance properties we add are correctly set up for all construction paths.

dhirving committed:
Factor out ButlerConfig 'cls' logic

We need the logic for mapping the ButlerConfig 'cls' field to Butler type in both Butler() and the upcoming ButlerFactory, so move it to ButlerConfig.

dhirving committed:
Add ButlerFactory

Add a factory class for creating multiple Butler instances for multiple repositories in the repository index.  This is intended for use in long-lived multi-threaded services that will instantiate a new Butler for each end-user HTTP request.

dhirving committed:
Allow RemoteButler instances to share cache

We now inject cached data from RemoteButlerFactory into RemoteButler, so that multiple RemoteButler instances connected to the same server can share a cache.

dhirving committed:
Fix docker smoke test race condition

With the changes to Butler construction, creating a RemoteButler in the Docker smoke test is now faster.  This was causing the Docker smoke test to be unreliable, because the test was sometimes trying to connect before the server had booted up.

dhirving committed:
Add towncrier

timj commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -297,17 +270,44 @@ def from_config(
+
+            factory = RemoteButlerFactory.create_factory_from_config(butler_config)
+            return factory.create_butler_with_credentials_from_environment(butler_options=options)
+
```
Can you add a `raise` here for clarity if there is no match to REMOTE or DIRECT? I know that there are currently only two values in the enum but this ties it up nicely. Consider using `match` here since that is designed for enums.

dhirving replied:
Because mypy's exhaustiveness checking will ensure that all the cases are handled, there's some value in not having the raise.  Without the raise, when someone adds an enum value, mypy will tell them all the places that need to do something with it.

I don't feel that strongly about it one way or the other though.

timj replied:
`match` naturally comes with a fall through clause that can raise. I feel better if the code looks like it's dealing with all the options without relying on mypy.

timj commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -1691,3 +1691,18 @@ def _query_dimension_records(
+    ) -> Butler:
+        """Return a new Butler instance connected to the same repository
+        as this one, but overriding ``collections``, ``run``,
+        ``inferDefaults``, and default data ID.
```
When this method becomes public I think it will trigger the numpydoc warning. Might want to add the public docs here now rather than later.

timj commented on python/lsst/daf/butler/_butler_config.py:
```diff
@@ -191,3 +194,20 @@ def __init__(
+            # missing the ``cls`` property.
+            return ButlerType.DIRECT
+        elif butler_class_name == "lsst.daf.butler.direct_butler.DirectButler":
+            return ButlerType.DIRECT
```
Can you please add a `cls` entry to one of the direct butler config files in `tests/config/basic` so we can trigger this code path?

timj commented on python/lsst/daf/butler/_butler_config.py:
```diff
@@ -191,3 +194,20 @@ def __init__(
+            return ButlerType.REMOTE
+        else:
+            raise ValueError(
+                f"Butler configuration requests to load unknown Butler class {butler_class_name}"
```
```suggestion
                f"Butler configuration requests to load unknown Butler class {butler_class_name!r}"
```
so the class name is quoted.

timj commented on python/lsst/daf/butler/_butler_factory.py:
```diff
@@ -0,0 +1,165 @@
+
+    Parameters
+    ----------
+    repositories : `Mapping`[`str`, `str`], optional
```
```suggestion
    repositories : `~collections.abc.Mapping`[`str`, `str`], optional
```

timj commented on python/lsst/daf/butler/_butler_factory.py:
```diff
@@ -0,0 +1,165 @@
+        config = ButlerConfig(config_uri)
+        butler_type = config.get_butler_type()
+
+        if butler_type == ButlerType.DIRECT:
```
Use `match` and catch if doesn't match?

timj commented on python/lsst/daf/butler/_butler_factory.py:
```diff
@@ -0,0 +1,165 @@
+        if self._repositories is None:
+            return ButlerRepoIndex.get_repo_uri(label)
+        else:
+            return self._repositories[label]
```
What happens if someone passes in a URI instead of a label? The `ButlerConfig` constructor tries to handle this. I'm not sure if that breaks here (or what the error message looks like if someone uses a label that isn't in the INDEX).

dhirving replied:
It just throws a KeyError for label not found.  It's important for security that this NOT work for URLs, because these labels are sometimes untrusted user input.

timj replied:
KeyError from python is a famously opaque error message.

timj commented on python/lsst/daf/butler/direct_butler.py:
```diff
@@ -109,119 +109,133 @@ class ButlerValidationError(ValidationError):
+        return self
+
+    @classmethod
+    def create_from_config(
```
We now have a `create_from_config` and a `from_config` ? (which I see registry factory also has). Is `create_from_config` pseudo private? Is it conceptually what `DirectButler.from_config` calls with cleaned up API?

dhirving replied:
I wanted to name this `from_config` but mypy wouldn't let me because it has a different signature than `Butler.from_config`.  There is no public `create_from_config` -- it's private because it's only on `DirectButler`.

timj replied:
It is public in the sense that:

```python
>>> b = Butler("dp02")
>>> help(b)
```
will show the method, and presumably Jupyter Notebooks will report the method. Maybe add an extra note to the docstring pointing to `Butler.from_config` as the general interface.

timj commented on python/lsst/daf/butler/direct_butler.py:
```diff
@@ -273,14 +287,11 @@ def _unpickle(
-        # MyPy doesn't recognize that the kwargs below are totally valid; it
-        # seems to think '**defaultDataId* is a _positional_ argument!
-        return cls(
+        return cls.create_from_config(
```
It occurs to me that pickling is not handling the without_datastore flag.

timj commented on python/lsst/daf/butler/direct_butler.py:
```diff
@@ -2369,3 +2382,5 @@ def _query_dimension_records(
     describe them (`StorageClassFactory`).
     """
+
+    _registry_shim: RegistryShim
```
Docstring for this and `_config`? The others do have docstrings even though private.

timj commented on python/lsst/daf/butler/remote_butler/_factory.py:
```diff
@@ -0,0 +1,111 @@
+    server_url : `str`
+        The URL of the Butler server that RemoteButler instances created by
+        this factory will connect to.
+
```
Remove blank line. Not needed.

timj commented on python/lsst/daf/butler/remote_butler/_remote_butler.py:
```diff
@@ -188,15 +153,19 @@ def isWriteable(self) -> bool:
         # Docstring inherited.
-        if self._dimensions is not None:
-            return self._dimensions
+        with self._cache.access() as cache:
```
Does this need the read lock? If the value `is not None` then no other thread is going to make it `None`.

dhirving replied:
I have not been able to locate any definitive information about Python's memory model for threading other than "if you use the synchronization primitives from the threading module then it works how you would expect".  

We probably don't need this on the read in the current version of CPython because of the GIL but that's not documented as permanent behavior anywhere that I am aware of.

timj commented on python/lsst/daf/butler/remote_butler/_remote_butler.py:
```diff
@@ -629,18 +598,21 @@ def _get_url(self, path: str, version: str = "v1") -> str:
             The full path to the endpoint.
         """
-        return f"{version}/{path}"
+        slash = "" if self._server_url.endswith("/") else "/"
```
Can we force `self._server_url` to always have the `/` so we don't need to check every time on get?

dhirving replied:
Yeah I had considered that but I was concerned someone might assign the server URL from somewhere else, forgetting the slash, and it would break intermittently.  By putting this check here we enforce it at the point where the code depends on it having a slash.

timj replied:
Isn't `_server_url` private?

dhirving replied:
Yes but RemoteButler is already almost 700 lines long and we haven't implemented 95% of its methods yet.  I don't see any reason not to enforce correctness at the point of use instead of adding a comment like "make extra sure this variable always has a slash at the end of it!"

timj replied:
I think I'm missing something. `_server_url` is set in one place in the constructor so my suggestion was to move this fixup line up to that point and normalize the string there. Then it's fixed once and never needs to be tested again.

I wasn't suggesting we document that all places that construct a RemoteButler must ensure the / is added.

dhirving replied:
Currently it is, but with the rate that these constructors are proliferating we may end up with it getting set more than one place.  

The only place that must have a slash is _get_url, so what is the benefit to doing this logic in the constructor?  The constructor is plenty busy with other issues already.

timj replied:
Ok. Leave it.  I don't really understand why we want to do a check and append every time the server is called rather than once when we construct the object. `_server_url` is entirely internal. We are in charge of how that variable is set.

timj commented on tests/test_butler_factory.py:
```diff
@@ -0,0 +1,73 @@
+
+    def test_factory_via_global_repository_index(self):
+        with tempfile.NamedTemporaryFile(suffix=".yaml") as index_file:
+            index_file.write(f"test_repo: {self.config_file_uri}\n".encode())
```
```suggestion
            print(f"test_repo: {self.config_file_uri}", file=index_file)
```
(and open temp file with `mode="w"`)? It is shorter and seems more natural.

timj commented on tests/test_thread_safe_cache.py:
```diff
@@ -0,0 +1,47 @@
+
+import unittest
+
+from lsst.daf.butler._utilities.thread_safe_cache import ThreadSafeCache
```
Maybe put the three thread safety tests in a single test file? They are very small.

timj reviewed (approved):
Looks really good.

timj commented on python/lsst/daf/butler/_butler_factory.py:
```diff
@@ -0,0 +1,165 @@
+
+
+class ButlerFactory:
+    """Factory for efficiently instantiating Butler instances from the
```
Now that I'm reading #931, we seem to be saying that server is required to use a label from a DAF_BUTLER_REPOSITORY_INDEX.  This makes me ponder that the name of the class is a bit too generic since DirectButler is not required to use the index labels and we don't expect this class to be used in non-server context.

dhirving replied:
Well, it's DAF_BUTLER_REPOSITORY_INDEX or "pass in whatever set of labels and config files you want."  I'd be open to renaming if you have any suggestions.

timj replied:
Since it has to be able to return DirectButler and RemoteButler for now the class can't really be ServerButlerFactory (as is the name of the variable you assign to). `LabeledButlerFactory`? `NamedButlerFactory`?

dhirving committed:
Rename ButlerFactory to LabeledButlerFactory

dhirving committed:
Explicitly handle unknown Butler types

Tim is not comfortable with relying on mypy and the existing exception in get_butler_type to ensure that all cases are handled.

dhirving committed:
Improve LabeledButlerFactory documentation

dhirving committed:
Improve DirectButler documentation

dhirving committed:
Consolidate threading utility tests into one file

andy-slac commented on python/lsst/daf/butler/_utilities/thread_safe_cache.py:
```diff
@@ -0,0 +1,78 @@
+    """
+
+    def __init__(self) -> None:
+        # This mutex may not actually be necessary for the current version of
+        # CPython, because both the 'get' and 'set' operation use underlying C
+        # functions that don't drop the GIL (at first glance anyway.) This is
+        # not documented anywhere that I can find, however, so better safe than
+        # sorry.
```
Extremely late comment - all built-in Python types are guaranteed to be thread-safe in CPython, there is a comment about it in https://docs.python.org/3/glossary.html#term-global-interpreter-lock. But an explicit mutex may be useful once we switch to GIL-free mode 😀
